fix #6039 fix(nimbus): Display warning for polling error, allow query refetch with useRefetchOnError hook

LZoog opened the pull request:
fixes #6039 

Because:
* When a polling error occurs, users are seeing the Apollo error alert on the entire page, and showing an alert is preferred since this error can happen when Experimenter is in mid-deploy
* Users are also sometimes seeing an error on the Home page that goes away after refresh, likely also due to Experimenter mid-deploy

This commit:
* Checks in AppLayoutWithExperiment for an apollo error after experiment data is present, signalling a polling error, and displays a warning-style alert above the rest of the page rather than only an alert. Polling is retried after the pollInterval passes.
* Adds useRefetchOnError hook and implements it on PageHome and App to cover getAllExperiments and the config query. This displays a warning-style alert informing the user the query will be retries, retries the query after a settimeout (5 seconds), and returns the Apollo error alert if an error occurs again
* Refactors AppLayoutWithExperiment stories and gives the component a default testId, adds tests for both pages

LZoog commented on app/experimenter/nimbus-ui/src/components/AppLayoutWithExperiment/index.stories.tsx:
```diff
@@ -72,4 +72,24 @@ storiesOf("components/AppLayoutWithExperiment", module)
+          )}
+        </AppLayoutWithExperiment>
+      </RouterSlugProvider>
+    );
```
Hmm guess I could update this file to the new Storybook syntax, will do

LZoog commented:
Alright, I've refactored the storybook file to reflect the new syntax.

The ticket suggests a try/catch and retry after X number of seconds, but the error from polling just comes back as `error` from `useQuery` and I tried some things that made this PR a lot more complicated that I was having trouble getting working.

This just displays an alert to the user and states polling will be retried to whatever we have set on the `pollInterval`, which defaults to 30 seconds.

See Storybook: https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/2f0c6066eee306df3967568f867cfc25e98c0de4/nimbus-ui/index.html?path=/story/components-applayoutwithexperiment--polling-error

LZoog commented:
Oh _man_, in checking out the ticket again, I just realized this was first reported in the **directory** view and secondary in the summary page. 😩

Right now, this only covers the Summary page. I'm going to revert this to a draft and keep hacking away at it to cover both pages.

jaredlockhart commented:
@LZoog does it poll on the directory view?

LZoog commented:
@jaredlockhart we don't poll on that page, but we can try to execute the query again after X number of seconds if it errors. From that Slack thread it seems like an immediate refresh tends to fix the problem, so maybe we try requerying after 5 seconds and if it still errors, just show the Apollo GQL error.

LZoog commented on app/experimenter/nimbus-ui/src/components/PageHome/index.stories.tsx:
```diff
@@ -41,6 +41,20 @@ export const Basic = storyTemplate([mockDirectoryExperimentsQuery()]);
+    error: new Error("boop, something's actually wrong"),
+  };
+  return (
+    <MockedCache mocks={[mockWithError, mockWithError]}>
```
FWIW I think we have room for a little cleanup here and I noticed the "Loading" story hasn't worked since we started handling the errors; filed #6113 as a follow up maintenance task.

LZoog commented:
Okay, I made the modifications on `PageHome` and fixed several merge conflicts that emerged. See my updated commit message/comment for details.

Storybooks of interest:
* [AppLayoutWithExperiment](https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/7508e50ea8d1ee87f00e4d0f74fe628cae8bf146/nimbus-ui/index.html?path=/story/components-applayoutwithexperiment--polling-error) polling error
* [PageHome](https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/7508e50ea8d1ee87f00e4d0f74fe628cae8bf146/nimbus-ui/index.html?path=/story/pages-home--query-error) getAllExperiments error/retry + error fallback

P.S. normally I hate disappearing messages that users don't have a chance to read if they wanted, but I didn't think we should let this hang out for 10-15 seconds and with Experimenter being an internal tool and this being a weird edge case 🤷‍♀️

LZoog commented on app/experimenter/nimbus-ui/src/components/PageHome/index.tsx:
```diff
@@ -40,6 +43,13 @@ export const Body = () => {
+        setHasRefetched(true);
+      }, REFETCH_DELAY);
+      return <RefetchAlert />;
+    }
```
A few things probably worth calling out:

* This does not catch and retry for JSON.parse-errors only and instead retries if anything fails, but I could change the conditional on L46 to something like `!hasRefetched && error.message.includes("JSON.parse")` if preferred, but not sure if there's ever any other kind error that occurs mid-deploy
* Errors should be captured by Sentry in `ApolloErrorAlert` (if the refetch fails), but not if something fails just once. It didn't seem like we wanted to capture it if it went away after refetch since it's likely a false positive.
* We're also not sending the polling-specific errors to Sentry since it feels edge casey and shouldn't error after repoll, but can if we want to

jaredlockhart commented on app/experimenter/nimbus-ui/src/components/PageHome/index.tsx:
```diff
@@ -106,4 +116,17 @@ const PageHome: React.FunctionComponent<PageHomeProps> = () => {
 };
 
+const RefetchAlert = () => (
+  <Alert variant="warning" data-testid="refetch-alert" className="mt-4">
+    <Alert.Heading>Error fetching experiments</Alert.Heading>
+    <p>
+      An error occured. This usually happens when Experimenter is mid-deploy.
+    </p>
+    <p>
+      Fetching experiments will be retried one time automatically in{" "}
+      {REFETCH_DELAY / 1000} seconds.
+    </p>
+  </Alert>
```
Can we reuse this component?

LZoog replied:
Now that I'm going to add this to `App/index.tsx` we can. With the heading, body, and alert `variant` differing, I didn't share it with `ApolloErrorAlert` or `AppErrorBoundary` alert.

jaredlockhart commented on app/experimenter/nimbus-ui/src/components/PageHome/index.tsx:
```diff
@@ -40,6 +43,13 @@ export const Body = () => {
   }
 
   if (error) {
+    if (!hasRefetched) {
+      setTimeout(() => {
+        refetch();
+        setHasRefetched(true);
+      }, REFETCH_DELAY);
+      return <RefetchAlert />;
```
So we use this logic for fetching experiments but not configuration on the home page?  What's the reasoning there?

LZoog replied:
Ohh, good call out. I actually just forgot that I need to add this to `App/index.tsx` as well to cover both queries, will add it there.

jaredlockhart reviewed (commented):
Tested locally, looks great, works great, just curious how we should handle errors on the initial load of the landing page?

jodyheavener reviewed (commented):
Just one thought!

LZoog commented on app/experimenter/nimbus-ui/src/components/PageHome/index.tsx:
```diff
@@ -40,6 +43,13 @@ export const Body = () => {
 
   if (error) {
+    if (!hasRefetched) {
+      setTimeout(() => {
```
Oh funny, I didn't see your comment but just pushed something similar since I saw the memory leak while testing.

LZoog commented:
Pushed another update - I've created a new hook, `useRefetchOnError`, since we needed to refetch the config as well as getAllExperimentsQuery, and added two stories for it.

I didn't end up reusing `RefetchAlert` with a polling error since the entire message heading/body differs and the `if (error) { return ErrorAlert }` bits wouldn't be as clean/simple. I also used a more general refetch message so it can be used for the config and all experiments query error.

I can remove any tests that might seem superfluous, I left the ones I wrote before this push except for one I believe that I moved into the hook test, but it seemed we should double check it works as expected on the pages we're using it, and we need the line coverage.

* [hook story](https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/6335fe007812efd1d3880f4bfc27656e24bdc09d/nimbus-ui/index.html?path=/story/hooks-userefetchonerror--error-and-refetch-error)
* [home page story](https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/6335fe007812efd1d3880f4bfc27656e24bdc09d/nimbus-ui/index.html?path=/story/pages-home--query-error)
* [polling story](https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/6335fe007812efd1d3880f4bfc27656e24bdc09d/nimbus-ui/index.html?path=/story/components-applayoutwithexperiment--polling-error

jodyheavener reviewed (approved):
LGTM!
